DB Information Schema - COLUMNS

jean-bernard-valentaten opened the pull request:
## Problem
This is a subset of #264 , covering only COLUMNS and TABLES interfaces.

Different RDBMS return Information Schema (metadata) information in different format. Magento framework should allow to switch between different implementations in order to support those different versions of DBs, as well as differences between versions.

## Solution
Introduce granular interfaces, based in Information Schema standard (looking into MySQL and MariaDB definitions in this scope, links are in the document). Application can switch between different implementation based on the profile.

larsroettig commented:
@akaplya can you review it pls

larsroettig commented:
Kann be merged told me @akaplya

melnikovi commented on design-documents/database/declarative-schema-database-compatibility.md:
````diff
@@ -0,0 +1,1093 @@
+```
+
+```
+class InformationSchema\Table\VarBitColumn extends Column
````
Do we need to introduce types that are not used by application currently? We will have to make sure that these types support across different databases engines.

jean-bernard-valentaten replied:
@melnikovi No they are not necessarily needed, I just added them for the sake of being complete. Technically both `bit` and `bit varying` have been dropped as of SQL-2003, so we can stimply ommit these.

melnikovi replied:
Yes, let's omit them.

jean-bernard-valentaten replied:
Ok, I removed the classes for `bit` and `bit varying` from the proposal.

melnikovi commented on design-documents/database/declarative-schema-database-compatibility.md:
````diff
@@ -0,0 +1,1093 @@
+    public function getDefaultValue(): ?float {}
+}
+```
+Note that ANSI-SQL defines the exact numerical data types `DECIMAL` and `NUMERIC` as being equivalent. MySQL / MariaDB
````
I would try to make columns types generic and delegate responsibility of mapping them to particular database engine to management classes.

jean-bernard-valentaten replied:
@melnikovi Agreed

melnikovi replied:
It's ok to not include this in the design. The idea is that providers will map database specific types to generic types in application.

jean-bernard-valentaten replied:
I added a comment in the section "Providers" stating what we have discussed.

melnikovi commented on design-documents/database/declarative-schema-database-compatibility.md:
```diff
@@ -0,0 +1,1093 @@
+    /**
+     * Returns the character set name.
+     *
+     * Note that not all RDBMS support this, returning `null` indicates whether this is supported.
```
Then we can't make type like this.

jean-bernard-valentaten replied:
@melnikovi Do you suggest to drop all non-ANSI-SQL features from column classes or do you have something else in mind?

melnikovi replied:
Yes, let's limit column types to currently used (can check declarative schema to see what is used).

melnikovi commented on design-documents/database/declarative-schema-database-compatibility.md:
````diff
@@ -0,0 +1,1093 @@
+
+```
+information_schema_profiles:
+  mysql-5.6:
````
Are we going to have these values in env.php?

larsroettig commented:
@melnikovi in PHP we do not have **enum** design pattern to archive this we need to use a library like https://github.com/spatie/enum or https://docs.spatie.be/enum/v2/introduction/.

melnikovi commented:
@larsroettig I was referring to this https://www.php.net/manual/en/class.splenum.php.

jean-bernard-valentaten commented:
@melnikovi I have a couple of questions about you considerations:

> Introducing enums for collation and charset

Fine with me, but bear in mind that these vary from RDBMS to RDBMS, thus those enums would have to be located in the DBAL.

> What should be return type of getDefaultValue?

Can you elaborate what you mean exactly? Imho the return type of `getDefaultValue` should be tied to the type of column but if you want it to be a generic type, I'd suggest making it `?string`. Is that what you are trying to say?

> What should be return type of getOnUpdate?

What do you propose?

jean-bernard-valentaten commented:
> Introduce enum for engine. Need to think what MySQL engines do we need to support and create enum based on that

I added that to the proposal. We may not be able to use `SplEnum` though as it is contained in a PECL package only.

> Need to remove TYPE from column classes

I removed them as requested

> Remove InformationSchema\Table\IntColumn:: getType and create different classes instead

That was already done, there are type specific classes for all ANSI-SQl variants of `integer`.

melnikovi commented:
Open questions will be finalized in subsequent proposals.
